[Feature] Implement /changename & related script bindings. Clean up #set name

catapultam-habeo opened the pull request:
# Description

Implements #4260. 

* Implements the following script bindings;
`Client::GrantNameChange()` -> Immediately invokes the Name Change window, and allows usability of the manually invoked /changename command. Name chosen is run through the character creation name filters, the vanity pet name filters, and the profanity filters and rejected as appropriate. (A client sharing a name with an NPC seems to cause clientside memory corruption). While usability is enabled, the window will be presented upon initial login for the character, and dismissing it with the 'Don't Remind Me Again' button will cause this behavior to be suppressed.
`Client::IsNameChangeAllowed()` -> Returns true if name change is currently enabled, false if disabled.
`Client::ClearNameChange()` -> Removes pending name change usability. Returns true if succeeded, false if there was no usability permitted.

* Implements the rule `Character::AlwaysAllowNameChange` which allows /changename to always work without needing to be enabled by `Client::GrantNameChange()`.

* Refactors the `#set name` GM-command to use the same code paths as the user-facing system. The notable improvement here is that it no longer requires the player to be kicked to character select.

* Refactors `quest::rename(...)` to use the same code path as the user-facing system. Likewise no longer requires the player to be kicked to character select.

* Deprecates `Client::ChangeFirstName(in_firstname, gmname)`, as the `OP_GMNameChange` no longer functions in ROF2 as it apparently did in the past. There is more room here to refine the handler for this packet, as the only way this packet is processed is as this struct in RoF2;
```
struct AltChangeName_Struct 
{
  /*00*/ char new_name[64];
  /*40*/ char old_name[64];
  /*80*/ int  response_code;
};
```

## Type of change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)

# Testing

```
-- #Arias.lua
function event_say(e)
 e.self:Say("Name Change is Currently : " ..tostring(e.other:IsNameChangeAllowed()))
 e.other:GrantNameChange()
 e.self:Say("Name Change is now... :" ..tostring(e.other:IsNameChangeAllowed()))
end
```
![image](https://github.com/user-attachments/assets/7731297e-d0ca-4cfa-b11d-31df34fb48ff)

![image](https://github.com/user-attachments/assets/09faa2ed-7c35-4edf-a8ba-28f4e6a816d3)
![image](https://github.com/user-attachments/assets/97297ed7-178c-4691-937f-d9d109314963)
![image](https://github.com/user-attachments/assets/7ef22f7b-fa76-4435-a076-6176f774dee4)

Clients tested: 

RoF2 (required, older clients unsupported)

# Checklist

- [X] I have tested my changes
- [X] I have performed a self-review of my code. Ensuring variables, functions and methods are named in a human-readable way, comments are added only where naming of variables, functions and methods can't give enough context.
- [X] I own the changes of my code and take responsibility for the potential issues that occur

zimp-wow commented:
### Test Environment

Akk-stack running cata's branch
Base PEQ snapshot
Latest master branch of PEQ quests
Unmodified ROF2 client

### Test Case : Result

#set name
- No argument: Displays help usage
- Existing name: Error displayed about name already taken
- ' ' or " ": Error about name already taken
- Curse word: Error about name already taken
- Really long name: Error about name already taken
- Two words separated by space: Only first word is taken as the input
- Pet name in 'character_pet_name': Already taken
- Successful change: Character name updates in real-time

'/changename'
- Use without grant: Popup and chat message about not eligible.
- empty name: Submit button does nothing
- ' ' or " ": Input not allowed
- Curse word: Invalid name error popup and chat message
- Really long name: Input stopped at a max length
- Two words: Input not allowed
- Attempting second name change without re-grant: Not eligible
- Pet name: Already taken
- Successful change: Character name updates in real-time

API
- Lua:GrantChangeName: Working
- Lua:IsNameChangedAllowed: Working
- Lua:ClearNameChange: Working
- Lua:rename: Working, but does still say the 'kicking to character select' message.
- Perl:GrantChangeName: Working
- Perl:IsNameChangeAllowed: Working
- **Perl:ClearNameChange: Missing?**
- Rule: Working

### Issues/Observations

- The error message about the name being already taken appears to display stale data about the current name of the player.  Seems to require a relog or a zone change to update.  Seems very purely cosmetic and doesn't seem like it would be encountered in most use cases.
- GrantChangeName did not seem to open the window on the client which doesn't match the description in the PR.  Similarly the window does not appear on login.  Could only get it open with the /changename command.
- As mentioned about 'ClearNameChange' appeared to be missing in the perl api changes.

catapultam-habeo commented:
In reference to @zimp-wow test, I have added the missing `ClearNameChange` to perl api, adjusted `GrantNameChange` to be of type void instead of type bool (and allow re-initiating it even if the rule allows always name change or a credit is already issued).

The popup on login intentionally is not provoked when the rule to always allow changes is enabled.


I have attached some basic testing demos.

catapultam-habeo commented:
@Akkadius if you have any idea why the windows build failed here, I would appreciate the insight.

Akkadius commented:
> @Akkadius if you have any idea why the windows build failed here, I would appreciate the insight.

It's in the build output

```
C:\Windows\Temp\drone-dq8FW86mGv7FfgQQ\drone\src\zone\client.cpp(4809,2): error C2562: 'Client::GrantNameChange': 'void' function returning a value [C:\Windows\Temp\drone-dq8FW86mGv7FfgQQ\drone\src\win-build-x64\zone\zone.vcxproj]
C:\Windows\Temp\drone-dq8FW86mGv7FfgQQ\drone\src\zone\client.cpp(4800): message : see declaration of 'Client::GrantNameChange' [C:\Windows\Temp\drone-dq8FW86mGv7FfgQQ\drone\src\win-build-x64\zone\zone.vcxproj]
```

zimp-wow commented:
I wasn't seeing the windows pop up because my opcodes got patched.  Was testing on my older akk-stack install and it was probably still enabled from before I understood that its generally a bad thing to do.  So that clears up that bit of confusion.

Akkadius commented on zone/questmgr.cpp:
```diff
@@ -1292,15 +1292,14 @@ void QuestManager::rename(std::string name) {
 					name
 				).c_str()
 			);
-			initiator->Kick("Name was changed.");
```
Are you sure there aren't unexpected consequences of not issuing a kick ? We had this in here previously for a reason

catapultam-habeo replied:
Yes, I cleaned up the serverside ones, and the ROF2 client performs the updates needed between the packets issued as part of this PR.

catapultam-habeo replied:
Either way, if it turns out to be an issue, I'll fix it. I namechanged and went around and did a bunch of stuff and had no issues (zoning, combat, char state saves, etc etc), but we all know how wild the edge cases can be. Would you prefer this behavior be rule gated?

Akkadius commented on zone/perl_client.cpp:
```diff
@@ -3393,6 +3408,9 @@ void perl_register_client()
 	package.add("ChangeLastName", &Perl_Client_ChangeLastName);
 	package.add("GrantPetNameChange", &Perl_Client_GrantPetNameChange);
+	package.add("GrantNameChange", (void(*)(Client*))&Perl_Client_GrantNameChange);
+	package.add("IsNameChangeAllowed", (bool(*)(Client*))&Perl_Client_IsNameChangeAllowed);
```
We are trying to keep this list alpha ordered, I'd move IsNameChangeAllowed down

Akkadius commented on zone/lua_client.cpp:
```diff
@@ -3635,6 +3653,9 @@ luabind::scope lua_register_client() {
 	.def("ChangeLastName", (void(Lua_Client::*)(std::string))&Lua_Client::ChangeLastName)
 	.def("GrantPetNameChange", &Lua_Client::GrantPetNameChange)
+	.def("GrantNameChange", &Lua_Client::GrantNameChange)
+	.def("IsNameChangeAllowed", &Lua_Client::IsNameChangeAllowed)
```
Alpha order

Akkadius commented on zone/gm_commands/set/name.cpp:
```diff
@@ -24,17 +24,13 @@ void SetName(Client *c, const Seperator *sep)
 
-			c->Message(Chat::White, "Sending player to char select.");
-
-			t->Kick("Name was changed");
```
Same comment about not kicking and consequences

Akkadius commented on zone/client.cpp:
```diff
@@ -4737,6 +4756,57 @@ bool Client::KeyRingRemove(uint32 item_id)
+void Client::GrantNameChange() {
+
+	auto k = GetScopedBucketKeys();
+	k.key = "NameChangesAllowed";
```
Let's keep bucket keys uniform to `snake_case` or slugged lower case

Akkadius commented on zone/client.cpp:
```diff
@@ -2579,10 +2568,40 @@ bool Client::ChangeFirstName(const char* in_firstname, const char* gmname)
+	return true;
+}
+
+bool Client::ChangeFirstName(const char* in_firstname)
```
Any reason we didn't just use std::string here ? We are using modern string types wherever possible

catapultam-habeo replied:
old method signature I think? want me to refactor to that?

Akkadius commented:
Few minor comments otherwise it looks great!

catapultam-habeo commented:
string handling revisions are limited in scope, I don't want to go down the rabbit hole of changing outside of scope atm

Akkadius reviewed (approved):
@zimp-wow is going to do one last round of testing

Akkadius commented on common/database.cpp:
```diff
@@ -955,6 +955,29 @@ bool Database::UpdateName(const std::string& old_name, const std::string& new_na
+	auto l = CharacterDataRepository::GetWhere(
+		*this,
+		fmt::format(
+			"`id` = '{}'",
```
Integers don't need to be quoted in queries but this will still work. It can sometimes throw off query planners though if you quote integers and can make slow queries
